Lib/unittest/test/testmock/testasync.py

```diff
@@ -528,11 +528,158 @@ class AsyncMockAssert(unittest.TestCase):
+
+    def test_assert_called_once_and_awaited_twice(self):
+        mock = AsyncMock(AsyncClass)
+        coroutine = mock.async_method()
+        mock.async_method.assert_called_once()
+        asyncio.run(self._await_coroutine(coroutine))
+        with self.assertRaises(RuntimeError):
+            # Cannot reuse already awaited coroutine
+            asyncio.run(self._await_coroutine(coroutine))
+        mock.async_method.assert_awaited()
+
+    def test_assert_awaited_but_not_called(self):
+        with self.assertRaises(AssertionError):
+            self.mock.assert_awaited()
+        with self.assertRaises(AssertionError):
+            self.mock.assert_called()
+        with self.assertRaises(TypeError):
+            # You cannot await an AsyncMock, it must be a coroutine
+            asyncio.run(self._await_coroutine(self.mock))
+
+        with self.assertRaises(AssertionError):
+            self.mock.assert_awaited()
+        with self.assertRaises(AssertionError):
+            self.mock.assert_called()
+
+    def test_assert_has_calls_not_awaits(self):
+        kalls = [call('NormalFoo')]
+        with self.assertWarns(RuntimeWarning):
+            self.mock('NormalFoo')
+        self.mock.assert_has_calls(kalls)
+        with self.assertRaises(AssertionError):
+            self.mock.assert_has_awaits(kalls)
+
+    def test_assert_has_mock_calls_on_async_mock(self):
+        mock = AsyncMock()
+        with self.assertWarns(RuntimeWarning):
+            mock()
+        kalls_empty = [('', (), {})]
+        self.assertEqual(mock.mock_calls, kalls_empty)
+
+        kalls_args = ([call('NormalFoo'), call('baz')])
+        with self.assertWarns(RuntimeWarning):
+            self.mock('NormalFoo')
+            self.mock('baz')
+        self.assertEqual(self.mock.mock_calls, kalls_args)
+
+        a_mock = AsyncMock(AsyncClass)
+        with self.assertWarns(RuntimeWarning):
+            a_mock.async_method()
+        self.assertEqual(a_mock.async_method.mock_calls, kalls_empty)
+        self.assertEqual(a_mock.mock_calls, [call.async_method()])
+
+        kalls_args_kwargs = [call(), call(1, 2, 3, a=4, b=5)]
+        with self.assertWarns(RuntimeWarning):
+            a_mock.async_method(1, 2, 3, a=4, b=5)
+        self.assertEqual(a_mock.async_method.mock_calls, kalls_args_kwargs)
+
+    def test_async_method_calls_recorded(self):
+        with self.assertWarns(RuntimeWarning):
+            self.mock.something(3, fish=None)
+            self.mock.something_else.something(6, cake=sentinel.Cake)
+
+        self.assertEqual(self.mock.something_else.method_calls,
+                        [("something", (6,), {'cake': sentinel.Cake})],
+                        "method calls not recorded correctly")
+        self.assertEqual(self.mock.method_calls, [
+            ("something", (3,), {'fish': None}),
+            ("something_else.something", (6,), {'cake': sentinel.Cake})
+        ],
+            "method calls not recorded correctly")
```

Comment:
Why you call `something` then `something_else.something`, but the tests check `something_else` first and `something`+`something_else_something` afterwards?

Reply:
Do you mean just swap the order of the asserts?

Reply:
Yes, it looked backwards and led me to wonder if there was something going on that required checking `something_else` before `something`.